### PR TITLE
Lazily initialize predicate for bound filter

### PR DESCRIPTION
**Pull request by navis**
Currently, bound filler makes new comparator for each value.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java [hunk 1]
```diff
@@ -34,35 +35,52 @@ public BoundFilter(final BoundDimFilter boundDimFilter)
     super(
         boundDimFilter.getDimension(), new Predicate<String>()
         {
+          private Predicate<String> predicate;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: volatile ?

**Comment by nishantmonu51** [comment]
:+1:

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java [hunk 2]
```diff
@@ -34,35 +35,52 @@ public BoundFilter(final BoundDimFilter boundDimFilter)
...
+            if (predicate == null) {
+              final Comparator<String> comparator;
+              if (boundDimFilter.isAlphaNumeric()) {
+                comparator = new AlphaNumericTopNMetricSpec(null).getComparator(null, null);
+              } else {
+                comparator = new LexicographicTopNMetricSpec(null).getComparator(null, null);
+              }
+              predicate = new Predicate<String>()
+              {
+                @Nullable
```
Comment (b-slim) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: i don't see how this return `null`, is this auto generated ?

**Comment by fjy** [comment]
@navis minor comments here, can we finish this up?

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java (hunk 1 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: done

File: processing/src/main/java/io/druid/segment/filter/BoundFilter.java (hunk 2 above)
Comment (navis) [inline processing/src/main/java/io/druid/segment/filter/BoundFilter.java]: removed

**Comment by b-slim** [comment]
:+1:

**Pull request merged**